## Review: Request changes — design is right, build is broken

**Verdict:** 🛑 Request changes. The architecture, dispatch fix, restore logic, re-entrancy guard, and test suite are all sound and solve the diagnosed bug cleanly. But the patch does not compile — CI has confirmed build failures on Linux and Windows. See specifics below.

### ❌ Build failure (blocking)

Both `build-linux` and `build-windows` fail at `games/oot/soh/GameExports_SingleExe.cpp`:

```
GameExports_SingleExe.cpp:176: error: 'gSaveContext' was not declared in this scope
GameExports_SingleExe.cpp:184: error: 'gSaveContext' was not declared in this scope
```

`integration-tests-linux` fails for the same reason (same TU compiled by that job).

**Cause.** The PR adds `Context_RestoreState(GAME_OOT, &gSaveContext, …)` and `gSaveContext.entranceIndex = targetEntrance` inside `OoT_Game_Resume` (near line 176–184 post-patch), but `extern \"C\" SaveContext gSaveContext;` in this TU is declared way down at line ~237, just above `Combo_CheckEntranceSwitch` where it was previously the only use site. The new use in `Game_Resume` is above the declaration, so the compiler rejects it.

**Fix.** Move `extern \"C\" SaveContext gSaveContext;` up into the top-of-file `extern \"C\"` block (alongside `gAudioContextInitalized` at ~line 32), or duplicate the declaration just before `OoT_Game_Resume`. The MM file doesn't hit this because it pulls in `global.h`, which declares `gSaveContext` for the MM layout.

### ✅ Dispatch fix — correct

`Combo_CheckEntranceSwitch` now resolves the source game via `Context_GetCurrentGame()` and calls `Combo_CheckCrossGameEntrance(gameId, …)` against the right entrance table. The early-boot `\"oot\"` default is safe given OoT always boots first, and `main.cpp` now keeps `Context_SetCurrentGame` in sync after `GameRunner_StartGame` and after every successful `GameRunner_SwitchTo`. Good.

One minor caveat worth calling out (the inline comment already does, but flagging for reviewers): when MM invokes this shared symbol, `sizeof(gSaveContext)` in the OoT TU still resolves to OoT's struct size. That's benign because `Context_FreezeState` clamps to the per-game N64 size internally, and the unified buffer is byte-addressed identically on both sides — but it means the `size` arg passed is effectively a lower bound, not ground truth. Fine as-is.

### ✅ SaveContext restore logic — correct

- `OoT_Game_Resume` restores only when `Context_HasFrozenState(GAME_OOT)` is true — first boot correctly skips.
- Startup-entrance preferred over frozen return-entrance matches how `Play_Init` consumes `Combo_GetStartupEntrance` on the next scene init; frozen fallback covers the mid-session resume case.
- Mirror fix in `MM_Game_Resume` writes to `gSaveContext.save.entrance` (MM's struct layout), which is the right field name for MM's `SaveContext`. Good.

### ✅ Re-entrancy guard — correct

`if (Combo_IsCrossGameSwitch()) return entranceIndex;` at the top of `Combo_CheckEntranceSwitch` prevents a second entrance-hook call on a subsequent transition frame from re-freezing on top of the already-captured state. This is the right place for the guard — earlier than the `Combo_CheckCrossGameEntrance` call that could otherwise mutate pending-switch state.

### ✅ Tests — solid, one small nit

The 4 new gtests in `ReturnPathTest` are a targeted regression suite for #170:
- `MMSideCheckFindsReturnEntrance` — covers the leg that silently no-op'd pre-fix.
- `OoTSideCheckDoesNotMatchMMEntrance` — negative guard against the old hardcoded-\"oot\" behavior.
- `FullRoundtripPreservesOoTSaveContext` — fingerprints both SaveContexts and asserts coexistence.
- `RestoreWithoutFreezeIsSafe` — edge case where `OoT_Game_Resume` reached without a prior freeze.

The `Test_Roundtrip` CTest now exercises `Combo_*` (the production C API) instead of the internal `Entrance_*` C++ API, which is the correct layer to lock in this regression.

**Nit:** The new tests hit `Combo_CheckCrossGameEntrance` directly rather than `Combo_CheckEntranceSwitch` (the fixed dispatch wrapper). That's reasonable — the dispatch logic is thin — but it means the `Context_GetCurrentGame()` branch in `GameExports_SingleExe.cpp` has no direct unit coverage; only the CTest `Roundtrip` indirectly validates it end-to-end. Fine for now, but worth a follow-up to parameterize the dispatch wrapper over `(currentGame, entranceId) → expectedTarget` if this code grows.

### Summary

Fix the `gSaveContext` declaration order and this is ready to ship. Design, restore logic, re-entrancy guard, and regression tests are all well-targeted for the root cause described.

--- a/combo/tests/context_test.cpp
+++ b/combo/tests/context_test.cpp
@@ -420,3 +420,113 @@ TEST(IntegrationTest, ReturnSwitchToMMRestoresFrozenState) {
     // Cleanup
     Context_ClearAllFrozenStates();
 }
+
+// ============================================================================
+// Return-path regression tests for issue #170
+// ============================================================================
+//
+// Before #170, Combo_CheckEntranceSwitch hardcoded "oot" as the source game
+// even when invoked from MM's z_play.c. That meant MM→OoT transitions never
+// resolved a matching cross-game link, and MM's SaveContext was never frozen.
+// These tests lock in the fix by exercising the C API that the game code
+// actually calls, asserting it works from both sides.
+
+TEST(ReturnPathTest, MMSideCheckFindsReturnEntrance) {
+    Context_Init();
+    Entrance_Init();
+    Entrance_RegisterTestLinks();
+
+    // Simulate MM stepping onto the return entrance (South Clock Town spawn 0).
+    // Using the C API with an explicit "mm" game id — this is the pathway the
+    // production fix takes via Context_GetCurrentGame() dispatch.
+    Combo_CheckCrossGameEntrance("mm", MM_ENTR_SOUTH_CLOCK_TOWN_0);
+
+    EXPECT_TRUE(Combo_IsCrossGameSwitch());
+    EXPECT_STREQ(Combo_GetSwitchTargetGameId(), "oot");
+    EXPECT_EQ(Combo_GetSwitchTargetEntrance(), OOT_ENTR_KOKIRI_FROM_MIDOS);
+
+    Entrance_ClearPendingSwitch();
+    Entrance_ClearLinks();
+}
+
+TEST(ReturnPathTest, OoTSideCheckDoesNotMatchMMEntrance) {
+    // Regression guard: the buggy implementation checked "oot" links for MM
+    // entrance 0xD800, which would silently match nothing. Verify that when
+    // the right game id ("oot") is paired with an MM entrance, no switch
+    // triggers — that's the correct negative behavior.
+    Context_Init();
+    Entrance_Init();
+    Entrance_RegisterTestLinks();
+
+    Combo_CheckCrossGameEntrance("oot", MM_ENTR_SOUTH_CLOCK_TOWN_0);
+    EXPECT_FALSE(Combo_IsCrossGameSwitch());
+
+    Entrance_ClearPendingSwitch();
+    Entrance_ClearLinks();
+}
+
+TEST(ReturnPathTest, FullRoundtripPreservesOoTSaveContext) {
+    Context_Init();
+    Entrance_Init();
+    Entrance_RegisterTestLinks();
+
+    // --- Leg 1: OoT → MM ---
+    // Populate OoT SaveContext with a recognizable fingerprint.
+    uint8_t ootSave[OOT_SAVE_CONTEXT_SIZE];
+    memset(ootSave, 0x11, sizeof(ootSave));
+    ootSave[0] = 'O'; ootSave[1] = 'o'; ootSave[2] = 'T'; ootSave[3] = '!';
+
+    // OoT's entrance hook fires while OoT is current.
+    Context_SetCurrentGame(GAME_OOT);
+    Combo_CheckCrossGameEntrance("oot", OOT_ENTR_MIDOS_HOUSE);
+    EXPECT_TRUE(Combo_IsCrossGameSwitch());
+    uint16_t ootReturn = Combo_GetSwitchReturnEntrance();
+    Combo_FreezeState("oot", ootReturn, ootSave, sizeof(ootSave));
+    Entrance_ClearPendingSwitch();
+
+    // --- Leg 2: pretend to run MM and freeze MM state ---
+    Context_SetCurrentGame(GAME_MM);
+    uint8_t mmSave[MM_SAVE_CONTEXT_SIZE];
+    memset(mmSave, 0x22, sizeof(mmSave));
+    mmSave[0] = 'M'; mmSave[1] = 'M';
+
+    Combo_CheckCrossGameEntrance("mm", MM_ENTR_SOUTH_CLOCK_TOWN_0);
+    EXPECT_TRUE(Combo_IsCrossGameSwitch());
+    EXPECT_STREQ(Combo_GetSwitchTargetGameId(), "oot");
+    uint16_t mmReturn = Combo_GetSwitchReturnEntrance();
+    Combo_FreezeState("mm", mmReturn, mmSave, sizeof(mmSave));
+    Entrance_ClearPendingSwitch();
+
+    // --- Leg 3: restore OoT on return, verify data integrity ---
+    uint8_t restoredOoT[OOT_SAVE_CONTEXT_SIZE] = {0};
+    int result = Combo_RestoreState("oot", restoredOoT, sizeof(restoredOoT));
+    EXPECT_EQ(result, 1);
+    EXPECT_EQ(restoredOoT[0], 'O');
+    EXPECT_EQ(restoredOoT[1], 'o');
+    EXPECT_EQ(restoredOoT[2], 'T');
+    EXPECT_EQ(restoredOoT[3], '!');
+    EXPECT_EQ(Combo_GetFrozenReturnEntrance("oot"), OOT_ENTR_KOKIRI_FROM_MIDOS);
+
+    // MM's frozen state still sits alongside OoT's — both coexist.
+    uint8_t restoredMM[MM_SAVE_CONTEXT_SIZE] = {0};
+    EXPECT_EQ(Combo_RestoreState("mm", restoredMM, sizeof(restoredMM)), 1);
+    EXPECT_EQ(restoredMM[0], 'M');
+
+    // Cleanup
+    Context_ClearAllFrozenStates();
+    Entrance_ClearLinks();
+}
+
+TEST(ReturnPathTest, RestoreWithoutFreezeIsSafe) {
+    // Edge case: first boot of a game — OoT_Game_Resume could be reached
+    // without any prior freeze (e.g. if GameRunner misbehaves). The restore
+    // must refuse rather than memcpy zeros over a live SaveContext.
+    Context_Init();
+    Context_ClearAllFrozenStates();
+
+    uint8_t buffer[OOT_SAVE_CONTEXT_SIZE];
+    memset(buffer, 0xFF, sizeof(buffer));
+    int result = Context_RestoreState(GAME_OOT, buffer, sizeof(buffer));
+    EXPECT_EQ(result, 0);
+    EXPECT_EQ(buffer[0], 0xFF);  // Untouched
+}

--- a/games/mm/2s2h/GameExports_SingleExe.cpp
+++ b/games/mm/2s2h/GameExports_SingleExe.cpp
@@ -21,6 +21,8 @@
 
 #include "game_lifecycle.h"
 #include "integration_test_hooks.h"
+#include "context.h"
+#include "entrance.h"
 #include <ship/resource/ResourceManager.h>
 #include <ship/resource/ResourceLoader.h>
 #include <ship/resource/archive/ArchiveManager.h>
@@ -360,6 +362,29 @@ void MM_Game_Suspend(void) {
 void MM_Game_Resume(void) {
     fprintf(stderr, "[MM] Game_Resume called\n");
     fflush(stderr);
+
+    // Restore the frozen MM SaveContext captured before we left for OoT (#170).
+    // OoT may have scribbled over the unified gSaveContext storage while it
+    // was active (see src/common/unified_save.c), so we must re-hydrate MM's
+    // view on every resume. First boot of MM has no frozen state — in that
+    // case MM_InitFirstEntrySaveContext still handles bootstrap via #168.
+    if (Context_HasFrozenState(GAME_MM)) {
+        fprintf(stderr, "[MM] Restoring frozen SaveContext on resume\n");
+        fflush(stderr);
+        Context_RestoreState(GAME_MM, &gSaveContext, sizeof(gSaveContext));
+
+        // Prefer the startup entrance set for this switch; fall back to the
+        // return entrance recorded when we last froze MM.
+        uint16_t startup = Combo_GetStartupEntrance();
+        uint16_t returnEntrance = Context_GetFrozenReturnEntrance(GAME_MM);
+        uint16_t targetEntrance = startup != 0 ? startup : returnEntrance;
+        if (targetEntrance != 0) {
+            gSaveContext.save.entrance = targetEntrance;
+            fprintf(stderr, "[MM] Resume entrance: 0x%04X (startup=%u)\n",
+                    targetEntrance, startup != 0);
+        }
+    }
+
     // TODO: Restore MM audio, reload MM-specific resources if needed
 }
 

--- a/games/oot/soh/GameExports_SingleExe.cpp
+++ b/games/oot/soh/GameExports_SingleExe.cpp
@@ -33,6 +33,10 @@ extern "C" {
     void OoT_Audio_PreNMI(void);
     extern s32 gAudioContextInitalized;
     void Audio_InitMesgQueues(void);
+
+    // OoT's SaveContext (type defined via OTRGlobals.h -> z64save.h).
+    // Declared here so OoT_Game_Resume() can restore it on return from MM (#170).
+    extern SaveContext gSaveContext;
 }
 
 // Game state
@@ -256,9 +260,6 @@ extern "C" {
     bool Combo_IsGameSwitchRequested(void);
     void Combo_ClearGameSwitchRequest(void);
 }
-
-// OoT's SaveContext (type defined via OTRGlobals.h -> z64save.h)
-extern "C" SaveContext gSaveContext;
 
 static bool sLastF10State = false;
 

--- a/games/oot/soh/GameExports_SingleExe.cpp
+++ b/games/oot/soh/GameExports_SingleExe.cpp
@@ -15,6 +15,8 @@
 
 #include "game_lifecycle.h"
 #include "integration_test_hooks.h"
+#include "context.h"
+#include "entrance.h"
 #include "soh/Enhancements/game-interactor/GameInteractor.h"
 
 // External declarations from main.c and other C sources
@@ -156,12 +158,34 @@ void OoT_Game_Suspend(void) {
 }
 
 /**
- * Resume OoT after being suspended for a game switch (issue #160).
- * Reinitializes audio message queues for clean state.
+ * Resume OoT after being suspended for a game switch (issue #160, #170).
+ * - Restores frozen OoT SaveContext so gameplay state survives the MM
+ *   round-trip (MM scribbles over the unified gSaveContext storage while
+ *   it is active — see src/common/unified_save.c).
+ * - Reinitializes audio message queues for clean state.
  */
 void OoT_Game_Resume(void) {
     fprintf(stderr, "[OoT] Game_Resume called\n");
     fflush(stderr);
+
+    // Restore the frozen OoT SaveContext captured before we left for MM (#170).
+    // Only restores when a frozen state exists — first boot of OoT skips this.
+    if (Context_HasFrozenState(GAME_OOT)) {
+        fprintf(stderr, "[OoT] Restoring frozen SaveContext on resume\n");
+        fflush(stderr);
+        Context_RestoreState(GAME_OOT, &gSaveContext, sizeof(gSaveContext));
+
+        // Prefer an explicit startup entrance (set by main.cpp for this
+        // switch); fall back to the return entrance recorded at freeze time.
+        uint16_t startup = Combo_GetStartupEntrance();
+        uint16_t returnEntrance = Context_GetFrozenReturnEntrance(GAME_OOT);
+        uint16_t targetEntrance = startup != 0 ? startup : returnEntrance;
+        if (targetEntrance != 0) {
+            gSaveContext.entranceIndex = targetEntrance;
+            fprintf(stderr, "[OoT] Resume entrance: 0x%04X (startup=%u)\n",
+                    targetEntrance, startup != 0);
+        }
+    }
 
     // Reinitialize audio message queues for clean state (issue #160).
     // The audio context's queue pointers may be stale after suspend.
@@ -272,18 +296,42 @@ extern "C" bool Combo_CheckHotSwap(void) {
 
 /**
  * Check if an entrance triggers a cross-game switch.
- * Called from randomizer_entrance.c Entrance_OverrideNextIndex().
+ *
+ * Dispatches against the currently active game so both OoT→MM and MM→OoT
+ * work (issue #170). In single-exe mode this symbol is shared between both
+ * games' translation units; the MM code path calls into this exact function
+ * via the `Combo_CheckEntranceSwitch` extern, so we must freeze the right
+ * side's SaveContext depending on who's running.
+ *
+ * Called from OoT's z_play.c / randomizer_entrance.c and from MM's
+ * z_play.c / z_player.c.
  */
 extern "C" uint16_t Combo_CheckEntranceSwitch(uint16_t entranceIndex) {
-    // Check if this entrance triggers a cross-game switch
-    uint16_t result = Combo_CheckCrossGameEntrance("oot", entranceIndex);
-
-    // If a cross-game switch was triggered, freeze our state
+    // If a cross-game switch is already queued (e.g. re-entrant call from a
+    // subsequent transition frame), skip to avoid re-freezing with mutated
+    // state (death sequences, mid-transition saves) — the main loop will
+    // process the existing pending switch on the next iteration.
     if (Combo_IsCrossGameSwitch()) {
-        fprintf(stderr, "[COMBO] Cross-game switch! entrance=0x%04X\n", entranceIndex);
+        return entranceIndex;
+    }
+
+    // Resolve which game is running so we pick the correct entrance table
+    // and freeze the correct SaveContext interpretation. Default to OoT when
+    // the current-game tracker hasn't been populated yet (early boot).
+    GameId currentGame = Context_GetCurrentGame();
+    const char* gameId = (currentGame == GAME_MM) ? "mm" : "oot";
+
+    uint16_t result = Combo_CheckCrossGameEntrance(gameId, entranceIndex);
+
+    if (Combo_IsCrossGameSwitch()) {
+        fprintf(stderr, "[COMBO] Cross-game switch (%s)! entrance=0x%04X\n",
+                gameId, entranceIndex);
 
         uint16_t returnEntrance = Combo_GetSwitchReturnEntrance();
-        Combo_FreezeState("oot", returnEntrance, &gSaveContext, sizeof(gSaveContext));
+        // sizeof(gSaveContext) in this TU is OoT's SaveContext layout, but the
+        // underlying unified storage is identical regardless of caller and
+        // Context_FreezeState clamps to the per-game N64 size anyway.
+        Combo_FreezeState(gameId, returnEntrance, &gSaveContext, sizeof(gSaveContext));
         Combo_SignalReadyToSwitch();
     }
 

--- a/rsbs/src/main.cpp
+++ b/rsbs/src/main.cpp
@@ -366,6 +366,9 @@ int main(int argc, char** argv) {
         free(gameArgv);
         return 1;
     }
+    // Keep context's view of the current game in sync with the runner so that
+    // cross-game entrance hooks dispatch against the right game (issue #170).
+    Context_SetCurrentGame(selectedGame);
 
     // For MM integration tests, OoT is initialized first for Ship::Context,
     // then we switch to MM (which registers its own hooks in MM_Game_Init)
@@ -380,6 +383,7 @@ int main(int argc, char** argv) {
                 free(gameArgv);
                 return 1;
             }
+            Context_SetCurrentGame(GAME_MM);
         }
     }
 
@@ -438,6 +442,10 @@ int main(int argc, char** argv) {
             if (switchResult != 0) {
                 fprintf(stderr, "Error: Failed to switch to game (code %d)\n", switchResult);
                 keepRunning = false;
+            } else {
+                // Track the new active game so cross-game entrance hooks
+                // resolve to the correct side on the next round-trip (#170).
+                Context_SetCurrentGame(nextGame);
             }
         } else {
             // Normal exit

--- a/src/common/context.cpp
+++ b/src/common/context.cpp
@@ -7,6 +7,7 @@
  */
 
 #include "context.h"
+#include <cstdio>
 #include <cstring>
 #include <vector>
 #include <algorithm>
@@ -232,7 +233,20 @@ bool Context_HasPendingSwitch(void) {
     return ComboContext_IsSwitchPending();
 }
 
-// Note: Context_ProcessSwitch() is implemented in switch.cpp
+void Context_SetCurrentGame(GameId game) {
+    gCurrentGame = game;
+    fprintf(stderr, "[Context] Current game set to %s\n", Game_ToString(game));
+}
+
+GameId Context_GetCurrentGame(void) {
+    return gCurrentGame;
+}
+
+// Note: Context_ProcessSwitch() and Context_IsSwitchInProgress() are implemented
+// in switch.cpp. That translation unit is only linked when the legacy
+// OoT_/MM_FreezeState/ResumeFromContext symbols are also available, i.e. in
+// non-single-exe builds. Single-exe callers should not reference those
+// functions.
 
 // ============================================================================
 // C API implementation

--- a/src/common/switch.cpp
+++ b/src/common/switch.cpp
@@ -164,19 +164,9 @@ bool Context_IsSwitchInProgress(void) {
     return gSwitchInProgress;
 }
 
-/**
- * Set the current game (used during initialization)
- */
-void Context_SetCurrentGame(GameId game) {
-    gCurrentGame = game;
-    fprintf(stderr, "[Switch] Current game set to %s\n", Game_ToString(game));
-}
-
-/**
- * Get the current game
- */
-GameId Context_GetCurrentGame(void) {
-    return gCurrentGame;
-}
-
 } // extern "C"
+
+// Context_SetCurrentGame / Context_GetCurrentGame live in context.cpp so that
+// call sites in main.cpp / GameExports_SingleExe.cpp don't pull switch.cpp.o
+// (and its references to the legacy OoT_/MM_FreezeState/ResumeFromContext
+// symbols, which are only compiled in non-single-exe builds) into the link.

--- a/src/common/test_runner.cpp
+++ b/src/common/test_runner.cpp
@@ -116,51 +116,93 @@ TestResult Test_SwitchMMOoT(void) {
 }
 
 TestResult Test_Roundtrip(void) {
-    printf("[TEST] roundtrip: Full round-trip with state verification\n");
+    printf("[TEST] roundtrip: Full round-trip with state verification (issue #170)\n");
 
     // Initialize systems
     Context_InitFrozenStates();
+    Context_ClearAllFrozenStates();
     Entrance_Init();
     Entrance_RegisterDefaultLinks();
 
-    // Simulate OoT -> MM -> OoT round trip
-    // Step 1: OoT triggers switch to MM
-    Entrance_CheckCrossGame(GAME_OOT, OOT_ENTR_HAPPY_MASK_SHOP);
-    if (!Entrance_IsCrossGameSwitch()) {
-        printf("[TEST] FAIL: Step 1 - OoT switch not triggered\n");
+    // ------------------------------------------------------------------
+    // Leg 1: OoT -> MM via Happy Mask Shop
+    // ------------------------------------------------------------------
+    // Use the production C API (Combo_*) — this is the same surface the
+    // game code actually calls from z_play.c. Exercising it here guards
+    // against the pre-#170 bug where Combo_CheckEntranceSwitch hardcoded
+    // "oot" as the source game on both legs.
+    Combo_CheckCrossGameEntrance("oot", OOT_ENTR_HAPPY_MASK_SHOP);
+    if (!Combo_IsCrossGameSwitch()) {
+        printf("[TEST] FAIL: Leg 1 - OoT->MM switch not triggered\n");
+        return TEST_FAIL;
+    }
+    if (strcmp(Combo_GetSwitchTargetGameId(), "mm") != 0) {
+        printf("[TEST] FAIL: Leg 1 - Target should be mm, got %s\n",
+               Combo_GetSwitchTargetGameId());
         return TEST_FAIL;
     }
 
-    // Simulate freezing OoT state
+    // Freeze a fingerprint for OoT so we can verify integrity after the trip.
     uint8_t fakeOoTSave[OOT_SAVE_CONTEXT_SIZE] = {0};
     fakeOoTSave[0] = 0xDE;
     fakeOoTSave[1] = 0xAD;
-    Context_FreezeState(GAME_OOT, OOT_ENTR_MARKET_FROM_MASK_SHOP,
-                        fakeOoTSave, sizeof(fakeOoTSave));
-
+    fakeOoTSave[OOT_SAVE_CONTEXT_SIZE - 1] = 0xEF;  // Tail marker
+    Combo_FreezeState("oot", Combo_GetSwitchReturnEntrance(),
+                      fakeOoTSave, sizeof(fakeOoTSave));
     Entrance_ClearPendingSwitch();
 
-    // Step 2: MM triggers switch back to OoT
-    Entrance_CheckCrossGame(GAME_MM, MM_ENTR_SOUTH_CLOCK_TOWN_0);
-    if (!Entrance_IsCrossGameSwitch()) {
-        printf("[TEST] FAIL: Step 2 - MM switch not triggered\n");
+    // ------------------------------------------------------------------
+    // Leg 2: MM -> OoT via South Clock Town
+    // ------------------------------------------------------------------
+    // This is the leg that pre-#170 silently no-op'd because the shared
+    // Combo_CheckEntranceSwitch implementation looked up "oot" links for
+    // MM's entrance id 0xD800 and always missed.
+    Combo_CheckCrossGameEntrance("mm", MM_ENTR_SOUTH_CLOCK_TOWN_0);
+    if (!Combo_IsCrossGameSwitch()) {
+        printf("[TEST] FAIL: Leg 2 - MM->OoT switch not triggered\n");
+        return TEST_FAIL;
+    }
+    if (strcmp(Combo_GetSwitchTargetGameId(), "oot") != 0) {
+        printf("[TEST] FAIL: Leg 2 - Target should be oot, got %s\n",
+               Combo_GetSwitchTargetGameId());
         return TEST_FAIL;
     }
 
-    // Step 3: Verify OoT state can be restored
+    // Also freeze MM state; both games' frozen states must coexist.
+    uint8_t fakeMMSave[MM_SAVE_CONTEXT_SIZE] = {0};
+    fakeMMSave[0] = 0xBE;
+    fakeMMSave[1] = 0xEF;
+    Combo_FreezeState("mm", Combo_GetSwitchReturnEntrance(),
+                      fakeMMSave, sizeof(fakeMMSave));
+
+    // ------------------------------------------------------------------
+    // Leg 3: Verify OoT state can be restored after the round-trip.
+    // ------------------------------------------------------------------
     uint8_t restoredSave[OOT_SAVE_CONTEXT_SIZE] = {0};
-    if (!Context_RestoreState(GAME_OOT, restoredSave, sizeof(restoredSave))) {
-        printf("[TEST] FAIL: Step 3 - OoT state restore failed\n");
+    if (!Combo_RestoreState("oot", restoredSave, sizeof(restoredSave))) {
+        printf("[TEST] FAIL: Leg 3 - OoT state restore failed\n");
+        return TEST_FAIL;
+    }
+    if (restoredSave[0] != 0xDE || restoredSave[1] != 0xAD ||
+        restoredSave[OOT_SAVE_CONTEXT_SIZE - 1] != 0xEF) {
+        printf("[TEST] FAIL: Leg 3 - Restored OoT data corrupted\n");
         return TEST_FAIL;
     }
 
-    if (restoredSave[0] != 0xDE || restoredSave[1] != 0xAD) {
-        printf("[TEST] FAIL: Step 3 - Restored data mismatch\n");
+    // And MM's frozen state is still intact.
+    uint8_t restoredMM[MM_SAVE_CONTEXT_SIZE] = {0};
+    if (!Combo_RestoreState("mm", restoredMM, sizeof(restoredMM))) {
+        printf("[TEST] FAIL: Leg 3 - MM state restore failed\n");
+        return TEST_FAIL;
+    }
+    if (restoredMM[0] != 0xBE || restoredMM[1] != 0xEF) {
+        printf("[TEST] FAIL: Leg 3 - Restored MM data corrupted\n");
         return TEST_FAIL;
     }
 
-    printf("[TEST] PASS: Round-trip with state preservation verified\n");
+    printf("[TEST] PASS: Full OoT<->MM round-trip preserves both SaveContexts\n");
     Entrance_ClearPendingSwitch();
+    Context_ClearAllFrozenStates();
     return TEST_PASS;
 }
 

